specs/graph-protocol-hybrid-network/read-interface/README.md

```diff
@@ -1,357 +1,273 @@
- - `index`: `Object` - The [IndexRecord](#indexes) of the index being read from.
- - `op`: `String` - The name of the read operation.
- - `params`: `[any]` - The parameters passed into the called read operation.
+ - `query`: `Object` - The query to run against the given `subgraph` at the given `blockHash`.
 2. `Object` - A [Locked Transfer](../messages#locked-transfer) message which serves as a conditional micropayment for the read operation.
 
-The `readIndex` method returns the following:
-1. `Object`
- - `data`: `any` - The data retrieved by the read operation.
- - `attestation`: Object - An attestation that `data` is a correct response for the given read operation (see [Attestation](#attestation)).
+and returns an `Object` in the response with the following properties:
+1. `data`: `any` - The data retrieved by the query.
+1. `blockHash`: `Object` - The hashes of the blocks from which the data was queried in the same format as the `blockHash` in the request, except that the value `latest` will not be used.
+1. `attestation`: `Object` - An attestation that `data` is a correct response for the given read operation (see [Attestation](#attestation)).
 
+#### A Simple Example
 ```js
 // request
 {
-  "method": "readIndex",
+  "method": "query",
   "params": [
     {
-      "blockHash": "xbf133b670857b983fc1b8f08759bc860378179042a0dba30b30e26d6f7f919d1",
-      "subgraphID": "QmTeW79w7QQ6Npa3b1d5tANreCDxF2iDaAPsDvW6KtLmfB",
-      "index": {
-        "indexType": "kv"
+      "blockHash": {
+        "ethereum": "latest"
       },
-      "op": "get"
-      "params": ["User:1"]
+      "subgraphID": "QmRngUAWqpipGzJULkt6MC5aU1vJyxcbwnkNHeXmn3fHqh",
+      // TODO: Use AST, not query string
+      "query": "select id, name from domains where labelName = 'addr'"
     }
   ],
   "jsonrpc": "2.0"
 }
 // response
 {
   "data": {
-    "firstName": "Vitalik",
-    "lastName": "Buterin",
+    "domains": [
+      {
+        "id": "0x91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e2",
+        "name": "addr.reverse"
+      }
+    ]
+  },
+  "blockHash": {
+      "ethereum": "0xd2042f2b92fea0de6c4fca82e7d149b932693c4061421b31b159ec1be39e3477"
   },
   // TODO: Provide more realistic attestations
   "attestation": 0x0122340
 }
 ```
 
-##### Example - Entity doesn't exist
+## Type system
 
-```js
-// request
-{
-  "method": "readIndex",
-  "params": [
-    {
-      "blockHash": "xbf133b670857b983fc1b8f08759bc860378179042a0dba30b30e26d6f7f919d1",
-      "index": {
-        "indexType": "kv"
-      },
-      "op": "get"
-      "params": ["User:1"]
-    }
-  ],
-  "jsonrpc": "2.0"
-}
-// response
-{
-  "data": null,
-  // TODO: Provide more realistic attestations
-  "attestation": 0x0122340
-}
-```
+The precise set of values that needs to be supported still needs to be finalized, but the type system will need to at least allow for strings, ints, bigints, bigdecimals, and byte strings as primitive types.
```

Comment:
Let's make sure finalizing these types is tracked somewhere. Maybe link to that issue here.

Reply:
Tracked here: https://github.com/graphprotocol/graph-node/issues/1096